ddp-workspace/projects/ddp-dsm-ui/src/app/FON/components/paginator/paginator.component.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  ChangeDetectionStrategy,
+  Component,
+  EventEmitter,
+  Input,
+  OnChanges,
+  Output,
+  SimpleChanges
+} from '@angular/core';
+
+
+interface pageProps {
+  from: number;
+  to: number;
+}
+
+enum SHIFT_PAGE {
+  START = 'START',
+  END = 'END',
+  ONE_MORE = 'ONE_MORE',
+  ONE_LESS = 'ONE_LESS'
+}
+
+/* Defaults */
+const PAGE_SIZE_OPTIONS = [10, 25, 50];
+const ROWS_PER_PAGE = 10;
+const CURRENT_PAGE_INDEX = 1;
+const VISIBLE_PAGES = 3;
+
+
+@Component({
+  selector: 'app-paginator',
+  templateUrl: './paginator.component.html',
+  styleUrls: ['./paginator.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PaginatorComponent implements OnChanges {
+  @Input() totalCount: number;
+  @Input() rowsPerPage = 10;
+  @Input() currentPageIndex: number;
+  @Input() pageSizeOptions: number[];
+  @Input() visiblePages: number;
+
+  @Output() pageChanged: EventEmitter<pageProps> = new EventEmitter();
+
+  constructor() {}
+
+  ngOnChanges(changes: SimpleChanges): void {
+    this.setDefaultParams();
+    this.generatePagesArray();
+  }
+
+  public get currentPageRange(): (string|number)[] {
+    const currentRange = this.filteredCurrentPageRange;
+    this.fillWithDots(currentRange);
+    return currentRange;
+  }
+
+  public changePage(page: number): void {
+    if (typeof page === 'number') {
+      this.currentPageIndex = page;
+      this.pageChanged.emit(this.pageProps);
+    }
+  }
+
+  public shiftPage(order: string): void {
+    switch (order) {
+      case SHIFT_PAGE.START:
+        this.currentPageIndex = 1;
+        break;
+      case SHIFT_PAGE.END:
+        this.currentPageIndex = this.generatePagesArray().length;
+        break;
+      case SHIFT_PAGE.ONE_MORE:
+        this.currentPageIndex = Math.min(this.currentPageIndex + 1, this.generatePagesArray().length);
+        break;
+      case SHIFT_PAGE.ONE_LESS:
+        this.currentPageIndex = Math.max(this.currentPageIndex - 1, 1);
+        break;
+      default:
+        return;
+    }
+    this.pageChanged.emit(this.pageProps);
+  }
+
+  public setRows(rows: number): void {
+    this.rowsPerPage = rows;
+    this.changePage(1);
+    this.generatePagesArray();
+  }
+
+  /* Paginator Engine */
+
+  private setDefaultParams(): void {
+    if(!this.rowsPerPage || typeof this.rowsPerPage !== 'number')
```

Comment:
I believe you can skip this method. 
Just set the default values for the class properties, like:

```
export class PaginatorComponent implements OnChanges {
  @Input() rowsPerPage = DEFAULT_ROWS_PER_PAGE;
  ... etc.
```

Reply:
This method is not only setting the values but also checking if they are the right type. Let's leave it so for now, later I will try to improve by finding better solution.

Reply:
Oops, I missed the fact that it is called inside of ngOnChanges hook. 
Wondering in which cases you should reset the params to default ones?

Reply:
As an option you can try `@Input setters`

Reply:
I just read this [article](https://dev.to/angular/angular-setters-vs-ngonchanges-which-one-is-better-4f2b) as well. I think it's better to switch to setters :D . I will do it surely later. Thanks.